src/Microsoft.Security.Utilities.Core/SecretMasker.cs

```diff
@@ -171,177 +174,125 @@ public string MaskSecrets(string input, Action<Detection>? detectionAction = nul
-    /// </summary>
     public void AddLiteralEncoder(LiteralEncoder encoder)
     {
-        SecretLiteral[] originalSecrets;
-
-        // Read section.
+        SyncObject.EnterWriteLock();
         try
         {
-            SyncObject.EnterReadLock();
-
-            if (LiteralEncoders.Contains(encoder))
+            if (!_literalEncoders.Add(encoder))
             {
                 return;
             }
 
-            // Read the original value secrets.
-            originalSecrets = ExplicitlyAddedSecretLiterals.ToArray();
-        }
-        finally
-        {
-            if (SyncObject.IsReadLockHeld)
+            foreach (SecretLiteral originalSecret in _explicitlyAddedSecretLiterals)
             {
-                SyncObject.ExitReadLock();
+                string encodedValue = encoder(originalSecret.Value);
+                if (!string.IsNullOrEmpty(encodedValue))
+                {
+                    _encodedSecretLiterals.Add(new SecretLiteral(encodedValue));
+                }
             }
         }
-
-        // Compute the encoded values.
-        var encodedSecrets = new List<SecretLiteral>();
-        foreach (SecretLiteral originalSecret in originalSecrets)
+        finally
         {
-            string encodedValue = encoder(originalSecret.Value);
-            if (!string.IsNullOrEmpty(encodedValue) && encodedValue.Length >= MinimumSecretLength)
-            {
-                encodedSecrets.Add(new SecretLiteral(encodedValue));
-            }
+            SyncObject.ExitWriteLock();
         }
+    }
 
-        // Write section.
+    public IEnumerable<Detection> DetectSecrets(string input)
+    {
+        var stopwatch = Stopwatch.StartNew();
         try
         {
-            SyncObject.EnterWriteLock();
-
-            // Add the encoder.
-            _ = LiteralEncoders.Add(encoder);
-
-            // Add the values.
-            foreach (SecretLiteral encodedSecret in encodedSecrets)
-            {
-                _ = EncodedSecretLiterals.Add(encodedSecret);
-            }
+            return DetectSecretsCore(input);
         }
         finally
         {
-            if (SyncObject.IsWriteLockHeld)
-            {
-                SyncObject.ExitWriteLock();
-            }
+            Interlocked.Add(ref _elapsedMaskingTicks, stopwatch.ElapsedTicks);
         }
     }
 
-    public IEnumerable<Detection> DetectSecrets(string input)
+    private List<Detection> DetectSecretsCore(string input)
     {
-        if (string.IsNullOrEmpty(input))
+        if (input == null || input.Length < MinimumSecretLength)
         {
-            yield break;
+            return [];
         }
 
-        // Read section.
+        SyncObject.EnterReadLock();
         try
         {
-            SyncObject.EnterReadLock();
-            var stopwatch = Stopwatch.StartNew();
+            var detections = new List<Detection>();
```

Comment:
>List [](http://example.com/codeflow?start=33&length=4)

Gee, we sure have a lot of code within this lock. I look forward to talking to you about how we might improve this moving forward. Basically by controlling the configuration of regexes/literals/etc.

And then we only have the pesky minimum secret length thing to worry about. If it ever comes to it, though, maybe we could produce all the detections up front, and then only take the lock to apply the minimum secret length filter. 

Or you know what, just copy this value upfront, so that any multithreaded change around that value only is applied to any scans initiated *after* the change.